hilti/src/rt/types/stream.cc

```diff
@@ -69,53 +53,181 @@ void Chunk::trim(Offset o) {
     _offset = o;
 }
 
-SafeConstIterator View::find(Byte b, const SafeConstIterator& n) const {
-    for ( auto i = UnsafeConstIterator(n ? n : _begin); i != UnsafeConstIterator(end()); ++i ) {
+const Chunk* Chain::findChunk(Offset offset, const Chunk* hint_prev) const {
+    _ensureValid();
+
+    const Chunk* c = _head.get();
+
+    if ( hint_prev && hint_prev->offset() <= offset )
+        c = hint_prev;
+
+    while ( c && ! c->inRange(offset) )
+        c = c->next();
+
+    if ( c && ! c->inRange(offset) )
+        return nullptr;
+
+    return c;
+}
+
+Chunk* Chain::findChunk(Offset offset, Chunk* hint_prev) {
+    _ensureValid();
+
+    Chunk* c = _head.get();
+
+    if ( hint_prev && hint_prev->offset() <= offset )
+        c = hint_prev;
+
+    while ( c && ! c->inRange(offset) )
+        c = c->next();
+
+    if ( _tail && offset > _tail->endOffset() )
+        return _tail;
+
+    return c;
+}
+
+const Byte* Chain::data(Offset offset, Chunk* hint_prev) const {
+    auto c = findChunk(offset, hint_prev);
+    if ( ! c )
+        throw InvalidIterator("stream iterator outside of valid range");
+
+    return c->data(offset);
+}
+
+void Chain::append(std::unique_ptr<Chunk> chunk) {
+    _ensureValid();
+    _ensureMutable();
+
+    if ( _tail ) {
+        _tail->setNext(std::move(chunk));
+        _tail = _tail->last();
+    }
+    else {
+        assert(! _head);
+        chunk->setOffset(_head_offset);
+        chunk->setChain(this);
+        _head = std::move(chunk);
+        _tail = _head->last();
+    }
+}
```

Comment:
We are not at all invalidating `chunk` here even though this is not much different from any function taking a `Chain&&`. Explicitly invalidating allows callers to perform some operations on `move`'d from chunks, so if we _need to_ do it elsewhere I think we'd also need to do it here. Alternatively, make less promises for the behavior of `move`'d from `Chain` objects and invalidate nowhere.